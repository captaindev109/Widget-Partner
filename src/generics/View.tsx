import { ViewProps, EachComponentDataSchema } from '@/types/types';
import Card from './Card';

// TODO fix types
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const View = ({ children, components }: ViewProps): any => {
  return components.map((component: EachComponentDataSchema) => <Card key={component.id}>{children(component)}</Card>);
};

export default View;
