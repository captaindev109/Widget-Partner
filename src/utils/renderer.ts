import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
