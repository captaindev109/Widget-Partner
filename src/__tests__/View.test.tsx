/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import View from '@/generics/View';
import viewRenderer from '@/utils/renderer';

export const data = [
  {
    id: '1',
    component: 'Card',
    padding: 10,
    background: 'grey',
    children: [
      {
        id: '2',
        component: 'Image',
        src: 'https://picsum.photos/800/400',
        width: '800px',
        height: '400px',
      },
    ],
  },
];

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<HelloWorld /> Test Cases', () => {
  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<View components={data}>{(children) => viewRenderer(children, children.id)}</View>);
  });

  it('Snapshot test should pass', () => {
    const snap = renderer
      .create(<View components={data}>{(children) => viewRenderer(children, children.id)}</View>)
      .toJSON();
    expect(snap).toMatchSnapshot();
  });
});
