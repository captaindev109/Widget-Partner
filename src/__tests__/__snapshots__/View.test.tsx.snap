// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelloWorld /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 leCJCh"
>
  <div
    className="Card-sc-fb64gk-0 iqWWFd"
    id="1"
  >
    <img
      className="Image__StyledImage-sc-34jhra-0 HuVGL"
      height="400px"
      src="https://picsum.photos/800/400"
      width="800px"
    />
  </div>
</div>
`;
